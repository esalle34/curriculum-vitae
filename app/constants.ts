enum Pages {
  CV = "cv",
  COMMON = "common",
}

export { Pages };
