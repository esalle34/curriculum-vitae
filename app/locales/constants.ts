enum Languages {
  English = "en",
  Français = "fr",
  "Español" = "es",
}

enum Countries {
  France = "FR",
  USA = "US",
  "United Kingdom" = "GB",
  "España" = "ES",
  Luxembourg = "LU",
  Suisse = "CH",
  Belgique = "BE",
}

const languagesArray = Object.entries(Languages);
const countriesArray = Object.entries(Countries);

const locales = countriesArray.flatMap(([_countryKey, country]) =>
  languagesArray.map(
    ([_languageKey, language]): Locale => `${language}-${country}`
  )
);

export type Locale = `${Languages}-${Countries}`;
const defaultLocale: Locale = `${Languages.English}-${Countries.USA}`;

interface DecodeTranslationOptions {
  className?: string;
  htmlTag: string;
}

export interface DecodeTranslation {
  options: DecodeTranslationOptions;
  translation: string;
}

export {
  Languages,
  Countries,
  defaultLocale,
  languagesArray,
  countriesArray,
  locales,
};
