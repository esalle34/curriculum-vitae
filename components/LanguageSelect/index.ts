import CountrySelect from "./LanguageSelect";

export default CountrySelect;
