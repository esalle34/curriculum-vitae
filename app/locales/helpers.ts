import { Countries, DecodeTranslation, Languages, Locale } from "./constants";

const getDictionary = (language: Languages, page: string) => ({
  [language]: () =>
    import(`@/app/locales/${language}/${page}.json`).then(
      (module) => module.default
    ),
});

const getLanguageAndCountry = (locale: Locale) => {
  const [language, country] = locale.split("-");

  return { language: language as Languages, country: country as Countries };
};

const decodeTranslation = ({
  options: { htmlTag, className },
  translation,
}: DecodeTranslation) => {
  return translation
    .replace(`<0>`, `<${htmlTag} ${className ? `class="${className}"` : ""}>`)
    .replace(`</0>`, `</${htmlTag}>`);
};

export { getDictionary, getLanguageAndCountry, decodeTranslation };
