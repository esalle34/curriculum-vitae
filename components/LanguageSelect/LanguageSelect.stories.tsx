import type { Meta, StoryObj } from "@storybook/react";

import LanguageSelect from "./LanguageSelect";
import { Languages } from "@/app/locales/constants";

const meta: Meta<typeof LanguageSelect> = {
  component: LanguageSelect,
};

export default meta;

type Story = StoryObj<typeof LanguageSelect>;

/*
 *ðŸ‘‡ Render functions are a framework specific feature to allow you control on how the component renders.
 * See https://storybook.js.org/docs/api/csf
 * to learn how to use render functions.
 */
export const LanguageSelectFR: Story = {
  render: () => (
    <LanguageSelect language={Languages.FranÃ§ais} setNewLanguage={() => {}} />
  ),
};
