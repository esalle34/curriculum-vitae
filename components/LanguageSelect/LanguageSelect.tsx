import { Languages, languagesArray } from "@/app/locales/constants";
import { Dispatch, SetStateAction } from "react";

interface Props {
  language: Languages;
  setNewLanguage: Dispatch<SetStateAction<Languages>>;
}

const LanguageSelect = ({ language, setNewLanguage }: Props) => (
  <select
    className="my-2 p-2 border-2 border-slate-300 rounded"
    defaultValue={language}
    onChange={(e) => setNewLanguage(e.target.value as Languages)}
  >
    {languagesArray.map(([languageKey, languageValue]) => {
      return (
        <option key={languageValue} value={languageValue}>
          {languageKey}
        </option>
      );
    })}
  </select>
);

export default LanguageSelect;
