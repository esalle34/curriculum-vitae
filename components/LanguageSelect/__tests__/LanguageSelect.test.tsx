import { Languages } from "@/app/locales/constants";
import LanguageSelect from "..";
import { fireEvent, render } from "@testing-library/react";

describe("CountrySelect", () => {
  it("should render FR snapshot correctly", () => {
    const { container } = render(
      <LanguageSelect
        language={Languages.Français}
        setNewLanguage={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
  it("should render EN snapshot correctly", () => {
    const { container } = render(
      <LanguageSelect language={Languages.English} setNewLanguage={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });
  it("should render correctly with language select", () => {
    const setNewCountry = jest.fn();

    const { getByRole, getByText } = render(
      <LanguageSelect
        language={Languages.Français}
        setNewLanguage={setNewCountry}
      />
    );
    const select = getByRole("combobox");
    expect(select).toBeInTheDocument();
    fireEvent.change(select, { target: { value: "English" } });
    const optionUSA = getByText(`English`);
    expect(optionUSA).toBeInTheDocument();
    expect(setNewCountry).toHaveBeenCalled();
  });
});
