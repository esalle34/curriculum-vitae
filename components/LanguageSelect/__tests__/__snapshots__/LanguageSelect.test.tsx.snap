// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelect should render EN snapshot correctly 1`] = `
<div>
  <select
    class="my-2 p-2 border-2 border-slate-300 rounded"
  >
    <option
      selected=""
      value="en"
    >
      English
    </option>
    <option
      value="fr"
    >
      Français
    </option>
    <option
      value="es"
    >
      Español
    </option>
  </select>
</div>
`;

exports[`CountrySelect should render FR snapshot correctly 1`] = `
<div>
  <select
    class="my-2 p-2 border-2 border-slate-300 rounded"
  >
    <option
      value="en"
    >
      English
    </option>
    <option
      selected=""
      value="fr"
    >
      Français
    </option>
    <option
      value="es"
    >
      Español
    </option>
  </select>
</div>
`;
